pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionAssignmentTailer.java

```diff
@@ -18,56 +18,85 @@
 
-import lombok.extern.slf4j.Slf4j;
+import java.io.IOException;
 
 @Slf4j
-public class FunctionAssignmentTailer
-    implements java.util.function.Consumer<Message<byte[]>>, Function<Throwable, Void>, AutoCloseable {
+public class FunctionAssignmentTailer implements AutoCloseable {
 
     private final FunctionRuntimeManager functionRuntimeManager;
+    @Getter
     private final Reader<byte[]> reader;
-    private boolean closed = false;
+    private volatile boolean isRunning = false;
 
-    public FunctionAssignmentTailer(FunctionRuntimeManager functionRuntimeManager, Reader<byte[]> reader) {
+    private final Thread tailerThread;
+    
+    public FunctionAssignmentTailer(FunctionRuntimeManager functionRuntimeManager, ReaderBuilder readerBuilder, WorkerConfig workerConfig) throws PulsarClientException {
         this.functionRuntimeManager = functionRuntimeManager;
-        this.reader = reader;
-    }
+        
+        this.reader = readerBuilder
+          .subscriptionRolePrefix(workerConfig.getWorkerId() + "-function-runtime-manager")
+          .readerName(workerConfig.getWorkerId() + "-function-runtime-manager")
+          .topic(workerConfig.getFunctionAssignmentTopic())
+          .readCompacted(true)
+          .startMessageId(MessageId.earliest)
+          .create();
 
-    public void start() {
-        receiveOne();
+        this.tailerThread = new Thread(() -> {
+            while(isRunning) {
+                try {
+                    Message<byte[]> msg = reader.readNext();
+                    processAssignment(msg);
+                } catch (Exception e) {
+                    if (isRunning) {
+                        log.error("Encountered error in assignment tailer", e);
+
+                        // trigger fatal error
+                        // TODO add mechanism to notify main thread
+                    } else {
+                        if (!(e instanceof InterruptedException)) {
+                            log.warn("Encountered error when assignment tailer is not running", e);
+                        }
+                    }
+
+                }
+            }
+        });
+        this.tailerThread.setName("assignment-tailer-thread");
     }
 
-    private void receiveOne() {
-        reader.readNextAsync()
-                .thenAccept(this)
-                .exceptionally(this);
+    public void start() {
+        isRunning = true;
+        tailerThread.start();
     }
 
     @Override
     public void close() {
-        if (closed) {
-            return;
-        }
-        log.info("Stopping function state consumer");
+        log.info("Stopping function assignment tailer");
         try {
-            closed = true;
-            reader.close();
+            isRunning = false;
+            if (tailerThread != null && tailerThread.isAlive()) {
+                tailerThread.interrupt();
+            }
+            if (reader != null) {
+                reader.close();
+            }
```

Comment:
also one benefit of doing a join here would be that once the thread is done, we can make it null again.
This way we could actually do a start/close multiple times.

Reply:
Is there a use case for that right now?

Reply:
Doing this right now makes the code more complicated since the reader needs to be used in FuntcionRuntimeManager